snuba/datasets/errors_replacer.py

```diff
@@ -314,6 +314,32 @@ def process_exclude_groups_and_replacement_types_results(
     return needs_final, exclude_groups, replacement_types
 
 
+def get_latest_replacement_time_by_projects(
+    project_ids: Sequence[int], state_name: Optional[ReplacerState]
+) -> Optional[float]:
+    """
+    Given project ids, returns the highest timestamp for any group
+    exclude replacement in the projects.
+    """
+    p = redis_client.pipeline()
+
+    for project_id in project_ids:
+        key, _ = get_project_exclude_groups_key_and_type_key(project_id, state_name)
+        # Gets the item with the highest score for a key, along with the score itself
+        p.zrevrangebyscore(
+            key, float("inf"), float("-inf"), withscores=True, start=0, num=1
+        )
+    # results are in the form [[(str, float)], [(str, float)], ...]
+    results = p.execute()
```

Comment:
I don't think this is going to be enough to provide the effect you want.
If we add the condition to exclude groups to a query even if that does not overlap with the replacement, the performance issue is minimal with the only exception of queries that are demoted to FINAL because the number of groups to exclude is too high.

What we would like to avoid is to apply FINAL to a query that does not overlap with the replacement.
Which means the redis key you are interested into the is FINAL key (which will need to be provided with a timestamp) not the group exclusion key.

Were you planning to do this as an intermediate step? That would be fine. Otherwise remember there is an additional step to do.

Reply:
No, you will have to add a timestamp and keep it up to date to the FINAL case as well, to make it work.
If you rely on the TTL, your change won't have any effect as, when the TTL is passed, the key will be removed and FINAL will not be applied anyway.